code/game/objects/items/RCD.dm

```diff
@@ -61,59 +60,75 @@ RLD
 	silo_mats = null
 	return ..()
 
+/obj/item/construction/pre_attack(atom/target, mob/user, params)
+	if(istype(target, /obj/item/rcd_upgrade))
+		install_upgrade(target, user)
+		return TRUE
+	if(insert_matter(target, user))
+		return TRUE
+	return ..()
+
 /obj/item/construction/attackby(obj/item/W, mob/user, params)
-	if(iscyborg(user))
+	if(istype(W, /obj/item/rcd_upgrade))
+		install_upgrade(W, user)
+		return TRUE
+	if(insert_matter(W, user))
+		return TRUE
+	return ..()
+
+/// Installs an upgrade into the RCD checking if it is already installed, or if it is a banned upgrade
+/obj/item/construction/proc/install_upgrade(obj/item/rcd_upgrade/rcd_up, mob/user)
+	if(rcd_up.upgrade & upgrade)
+		to_chat(user, "<span class='warning'>[src] has already installed this upgrade!</span>")
 		return
-	var/loaded = 0
-	if(istype(W, /obj/item/rcd_ammo))
-		var/obj/item/rcd_ammo/R = W
+	if(rcd_up.upgrade & banned_upgrades)
+		to_chat(user, "<span class='warning'>[src] can't install this upgrade!</span>")
+		return
+	upgrade |= rcd_up.upgrade
+	if((rcd_up.upgrade & RCD_UPGRADE_SILO_LINK) && !silo_mats)
+		silo_mats = AddComponent(/datum/component/remote_materials, "RCD", FALSE, FALSE)
```

Comment:
/obj/item/rcd_upgrade really only has all the "diskettes" that upgrade the RCD.

I agree theres probably a better way to add them, but I know this is working, it will do for our purposes for now. I'd really like to get the airlock changes in.

In a future PR I will probably be changing it